### PR TITLE
feat: missing String.natcmp method

### PR DESCRIPTION
**Pull request by sclinede**

### CODE REVIEW

File: lib/string_tools/core_ext/string.rb [hunk 1]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
 
+  def self.natcmp(str1, str2)
+    str1, str2 = str1.dup, str2.dup
+    compare_expression = /^(\D*)((?:\d+(?:\.\d+)?)*)(.*)$/
+
+    # Remove all whitespace
+    str1.gsub!(/\s*/, '')
+    str2.gsub!(/\s*/, '')
+
+    while (str1.length > 0) or (str2.length > 0) do
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Do not use `do` with multi-line `while`.

File: lib/string_tools/core_ext/string.rb [hunk 2]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
+            when 1 then return 1
+          end
+        when -1 then return -1
+        when 1 then return 1
+      end # case
+
+    end # while
+
+    # Strings are naturally equal
+    return 0
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Redundant `return` detected.

File: lib/string_tools/core_ext/string.rb [hunk 3]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
+    str1.gsub!(/\s*/, '')
+    str2.gsub!(/\s*/, '')
+
+    while (str1.length > 0) or (str2.length > 0) do
+      # Extract non-digits, digits and rest of string
+      str1 =~ compare_expression
+      chars1, num1, str1 = $1.dup, $2.dup, $3.dup
+
+      str2 =~ compare_expression
+      chars2, num2, str2 = $1.dup, $2.dup, $3.dup
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Avoid the use of Perl-style backrefs.

File: lib/string_tools/core_ext/string.rb [hunk 4]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
+    compare_expression = /^(\D*)((?:\d+(?:\.\d+)?)*)(.*)$/
+
+    # Remove all whitespace
+    str1.gsub!(/\s*/, '')
+    str2.gsub!(/\s*/, '')
+
+    while (str1.length > 0) or (str2.length > 0) do
+      # Extract non-digits, digits and rest of string
+      str1 =~ compare_expression
+      chars1, num1, str1 = $1.dup, $2.dup, $3.dup
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Avoid the use of Perl-style backrefs.

File: lib/string_tools/core_ext/string.rb [hunk 5]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
+        when 0 # Non-digits are the same, compare the digits...
+          # If either number begins with a zero, then compare alphabetically,
+          # otherwise compare numerically
+          if !(num1[0] == 48 && num1[1] != 46) and !(num2[0] == 48 && num2[1] != 46)
+            num1, num2 = num1.to_f, num2.to_f
+          end
+
+          case (num1 <=> num2)
+            when -1 then return -1
+            when 1 then return 1
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Indent `when` as deep as `case`.

File: lib/string_tools/core_ext/string.rb [hunk 6]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
+      case (chars1 <=> chars2)
+        when 0 # Non-digits are the same, compare the digits...
+          # If either number begins with a zero, then compare alphabetically,
+          # otherwise compare numerically
+          if !(num1[0] == 48 && num1[1] != 46) and !(num2[0] == 48 && num2[1] != 46)
+            num1, num2 = num1.to_f, num2.to_f
+          end
+
+          case (num1 <=> num2)
+            when -1 then return -1
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Indent `when` as deep as `case`.

File: lib/string_tools/core_ext/string.rb [hunk 7]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
+            when 1 then return 1
+          end
+        when -1 then return -1
+        when 1 then return 1
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Indent `when` as deep as `case`.

File: lib/string_tools/core_ext/string.rb [hunk 8]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
+          # otherwise compare numerically
+          if !(num1[0] == 48 && num1[1] != 46) and !(num2[0] == 48 && num2[1] != 46)
+            num1, num2 = num1.to_f, num2.to_f
+          end
+
+          case (num1 <=> num2)
+            when -1 then return -1
+            when 1 then return 1
+          end
+        when -1 then return -1
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Indent `when` as deep as `case`.

File: lib/string_tools/core_ext/string.rb [hunk 9]
```diff
@@ -39,6 +39,44 @@ def naturalized
...
+      # Extract non-digits, digits and rest of string
+      str1 =~ compare_expression
+      chars1, num1, str1 = $1.dup, $2.dup, $3.dup
+
+      str2 =~ compare_expression
+      chars2, num2, str2 = $1.dup, $2.dup, $3.dup
+
+      # Compare the non-digits
+      case (chars1 <=> chars2)
+        when 0 # Non-digits are the same, compare the digits...
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Indent `when` as deep as `case`.

File: lib/string_tools/core_ext/string.rb [hunk 10]
```diff
@@ -39,6 +39,44 @@ def naturalized
     scan(/[^\d\.]+|[\d\.]+/).map{|f| f.match(/^\d+(\.\d+)?$/) ? f.to_f : f }
   end
 
+  def self.natcmp(str1, str2)
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Cyclomatic complexity for natcmp is too high. [12/6]<br>Assignment Branch Condition size for natcmp is too high. [35.3/35]<br>Perceived complexity for natcmp is too high. [9/7]

**Comment by cynek** [comment]
ok

**Comment by Napolskih** [comment]
ok

File: lib/string_tools/core_ext/string.rb [hunk 11]
```diff
@@ -84,4 +191,14 @@ def to_cp1251
...
   end
+
+  private
+
+  def surround_with_ansi(ascii_seq)
+    "#{ascii_seq}#{protect_escape_of(ascii_seq)}#{ANSI_CLEAR}"
+  end
+
+  def protect_escape_of(ascii_seq)
+    self.gsub(Regexp.new(Regexp.escape(ANSI_CLEAR)), "#{ANSI_CLEAR}#{ascii_seq}")
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Redundant `self` detected.

File: lib/string_tools/core_ext/string.rb [hunk 12]
```diff
@@ -39,12 +39,119 @@ def naturalized
...
+  # === Synopsys
+  #   Make text underlined (for ASCII terminals)
+  def underline
+    surround_with_ansi(ANSI_UNDERLINE)
+  end
+
+  # === Synopsys
+  #   remove colors from colorized string
+  def remove_colors
+    self.gsub(/\e\[\d+m/, '')
```
Comment (apress-hound) [inline lib/string_tools/core_ext/string.rb]: Redundant `self` detected.

**Comment by bibendi** [comment]
ok

**Comment by salteron** [comment]
Леня и :ok:

File: lib/string_tools/core_ext/string.rb (hunk 12 above)
Comment (sclinede) [inline lib/string_tools/core_ext/string.rb]: чего-то не уверен

**Comment by sclinede** [comment]
остальное не правлю. т.к. вынос кода.

**Pull request merged**